File: bsp/boot/arm/musca-b1/debug.c
/*-
 * Copyright (c) 2026, Champ Yen <champ.yen@gmail.com>
 * All rights reserved.
 */

#include <sys/param.h>
#include <boot.h>

#define UART_BASE CONFIG_PL011_PHY_BASE
#define UART_DR (*(volatile uint32_t*)(UART_BASE + 0x00))
#define UART_FR (*(volatile uint32_t*)(UART_BASE + 0x18))
#define FR_TXFF 0x20

void debug_putc(int c)
{
    while (UART_FR & FR_TXFF);
    UART_DR = c;
}

void debug_init(void)
{
}

File: bsp/boot/arm/musca-b1/debug.zig
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2026, Champ Yen <champ.yen@gmail.com>
// All rights reserved.

const ffi = @import("ffi");

const UART_BASE = ffi.cfg.CONFIG_PL011_PHY_BASE;

const UART_DR: *volatile u32 = @ptrFromInt(UART_BASE + 0x00);
const UART_FR: *volatile u32 = @ptrFromInt(UART_BASE + 0x18);

const FR_TXFF = 0x20;

pub fn debug_putc(c: c_int) void {
    while ((UART_FR.* & FR_TXFF) != 0) {}
    UART_DR.* = @intCast(@as(u8, @intCast(c)));
}

pub fn debug_init() void {
}
